src/main.rs

```diff
@@ -4,6 +4,216 @@
+use serde::Deserialize;
+use std::path::PathBuf;
+use std::time::SystemTime;
+use stream_rust_test::math::{BaseValue, MathFunction};
+use tokio::task::JoinSet;
+use tokio_stream::wrappers::ReadDirStream;
+use tokio_stream::StreamExt;
+use tracing::error;
+use tracing::log::debug;
+use tracing_subscriber::layer::SubscriberExt;
+use tracing_subscriber::util::SubscriberInitExt;
+use tracing_subscriber::EnvFilter;
+
+#[derive(Debug, Clone, Deserialize)]
+struct AstarteConfig {
+    /// Astarte realm
+    realm: String,
+    /// Device ID
+    device_id: String,
+    /// Device credential secret
+    #[serde(default)]
+    credentials_secret: Option<String>,
+    /// Device pairing token
+    #[serde(default)]
+    pairing_token: Option<String>,
+    /// Astarte pairing url
+    pairing_url: String,
+    /// Astarte store directory
+    store_directory: PathBuf,
+    /// Flag to ignore Astarte SSL errors
+    astarte_ignore_ssl: bool,
+    /// Path to folder containing the Astarte Device interfaces
+    interfaces_directory: PathBuf,
+}
+
+/// Configuration for the values to be sent to Astarte
+#[derive(Debug, Clone, Parser)]
+#[clap(version, about)]
+struct Config {
+    /// Math function the device will use to send data to Astarte
+    #[clap(short, long, default_value = "default", env = "MATH_FUNCTION")]
+    math_function: MathFunction,
+    /// Interface name to send data to
+    #[clap(
+        long,
+        default_value = "org.astarte-platform.genericsensors.Values",
+        env = "INTERFACE_NAME"
+    )]
+    interface_datastream_do: String,
+    /// Milliseconds the device must wait before sending data to Astarte
+    #[clap(short, long, default_value = "1000", env = "INTERVAL_BTW_SAMPLES")]
+    interval_btw_samples: u64,
+    /// Scale for the generation of the data to send
+    #[clap(short, long, default_value = "1.0", env = "SCALE")]
+    scale: f64,
+}
+
+#[tokio::main]
+async fn main() -> eyre::Result<()> {
+    color_eyre::install()?;
+
+    tracing_subscriber::registry()
+        .with(tracing_subscriber::fmt::layer())
+        .with(EnvFilter::from_default_env())
+        .try_init()?;
+
+    // time instant when the program starts its execution
+    let now = SystemTime::now();
+
+    // initialize CLI configuration options
+    let cli_cfg = Config::parse();
+
+    debug!("Parsed config: {:#?}", cli_cfg);
+
+    // Load astarte configuration
+    let astarte_cfg: AstarteConfig = load_astarte_cfg().await?;
+
+    debug!("Parsed Astarte config: {:#?}", astarte_cfg);
+
+    // define type of credential (pairing token or credential secret) to use to establish an MQTT
+    // connection with Astarte
+    let cred = if let Some(pairing) = astarte_cfg.pairing_token.as_deref() {
+        Credential::paring_token(pairing)
+    } else if let Some(secret) = astarte_cfg.credentials_secret.as_deref() {
+        Credential::secret(secret)
+    } else {
+        bail!("missing credential secret or pairing token");
+    };
```

Comment:
Invert the logic... from Readme `If both are present, the credential secret will be used`.
```suggestion
    let cred = if let Some(secret) = astarte_cfg.credentials_secret.as_deref() {
        Credential::secret(secret)
    }   else  if let Some(pairing) = astarte_cfg.pairing_token.as_deref() {
         Credential::paring_token(pairing)
    } else {
        bail!("missing credential secret or pairing token");
    };
```